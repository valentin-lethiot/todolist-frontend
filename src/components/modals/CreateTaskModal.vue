<script setup lang="ts">
import BaseModal from './BaseModal.vue'
import Input from '../forms/Input.vue'
import TextArea from '../forms/TextArea.vue'
import { computed, reactive, ref } from 'vue'
import { useTasksStore } from '@/stores/tasks.store'

const { createTask } = useTasksStore()

type Props = {
  isOpen: boolean
}
defineProps<Props>()

const emit = defineEmits(['close'])

const taskToCreate = reactive({
  title: '',
  description: '',
})

const isLoading = ref(false)

const isCreateButtonDisabled = computed(() => {
  return !taskToCreate.title || !taskToCreate.description
})

const resetForm = () => {
  taskToCreate.title = ''
  taskToCreate.description = ''
}

const validateForm = async () => {
  if (isCreateButtonDisabled.value) return

  isLoading.value = true
  await createTask({ ...taskToCreate })
  isLoading.value = false

  resetForm()
  emitClose()
}

const emitClose = () => emit('close')
</script>

<template>
  <BaseModal v-show="isOpen" :is-loading="isLoading">
    <div class="flex flex-col gap-4 w-full h-full">
      <div class="h-1/8 text-xl font-bold flex justify-between items-start">
        <p>Créer une tâche</p>
        <button
          @click="emitClose"
          class="rounded-full px-2 bg-gray-700 hover:bg-gray-600 text-lg font-bold text-white hover:cursor-pointer"
        >
          X
        </button>
      </div>

      <div class="h-6/8 flex flex-col gap-4">
        <Input label="Titre" v-model="taskToCreate.title" />
        <TextArea label="Description" v-model="taskToCreate.description" />
      </div>

      <div class="h-1/8 w-full flex gap-4 items-center justify-end">
        <button
          @click="emitClose"
          class="bg-gray-700 hover:bg-gray-600 rounded-lg text-lg font-bold p-2 text-white w-1/2 hover:cursor-pointer"
        >
          Annuler
        </button>
        <button
          class="bg-green-700 hover:bg-green-600 rounded-lg text-lg font-bold p-2 text-white w-1/2 hover:cursor-pointer disabled:bg-green-600/20 hover:disabled:cursor-not-allowed"
          :disabled="isCreateButtonDisabled"
          @click="validateForm"
        >
          Créer
        </button>
      </div>
    </div>
  </BaseModal>
</template>
