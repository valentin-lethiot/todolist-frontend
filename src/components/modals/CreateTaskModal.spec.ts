import { describe, beforeEach, expect, it } from 'vitest'
import CreateTaskModal from './CreateTaskModal.vue'
import { mount, VueWrapper } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'

describe('CreateTaskModal', () => {
  let wrapper: VueWrapper

  const defaultProps = {
    isOpen: true,
  }

  beforeEach(() => {
    setActivePinia(createPinia())
    wrapper = mount(CreateTaskModal, { props: defaultProps })
  })

  describe('render', () => {
    it('should render the modal', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('closing modal', () => {
    it('should emit close when clicking close button', async () => {
      await wrapper.find('#closeCross').trigger('click')
      expect(wrapper.emitted('close')).toBeTruthy()
    })

    it('should emit close when clicking cancel button', async () => {
      await wrapper.find('#cancelButton').trigger('click')
      expect(wrapper.emitted('close')).toBeTruthy()
    })
  })

  describe('validation button', () => {
    it('should be disabled when form is empty', () => {
      const createButton = wrapper.find('#validateButton')
      expect(createButton.attributes('disabled')).toBeDefined()
    })

    it('should be enabled when form is filled', async () => {
      await wrapper.find('input').setValue('Task Title')
      await wrapper.find('textarea').setValue('Task Description')

      const createButton = wrapper.find('#validateButton')
      expect(createButton.attributes('disabled')).toBeUndefined()
    })
  })
})
