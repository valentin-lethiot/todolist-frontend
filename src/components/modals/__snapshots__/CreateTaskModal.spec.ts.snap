// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateTaskModal > render > should render the modal 1`] = `
"<div id="create-task-modal" class="absolute w-screen h-screen top-0 left-0 bg-gray-900/40 z-50">
  <div class="w-full h-full flex flex-col justify-center items-center p-8">
    <div class="bg-gray-800 rounded-lg p-4 shadow-xl w-full max-h-2/3">
      <div class="flex flex-col gap-4 w-full h-full">
        <div class="h-1/8 text-xl font-bold flex justify-between items-start">
          <p>Créer une tâche</p><button id="closeCross" class="rounded-full px-2 bg-gray-700 hover:bg-gray-600 text-lg font-bold text-white hover:cursor-pointer"> X </button>
        </div>
        <div class="h-6/8 flex flex-col gap-4">
          <div><label class="block mb-1 text-md font-medium text-white">Titre</label><input type="text" class="bg-gray-50 border text-sm rounded-lg block w-full p-2 bg-gray-700 border-gray-600 text-white" required="" value=""></div>
          <div data-v-993744c5=""><label data-v-993744c5="" class="block mb-1 text-md font-medium text-white">Description</label><textarea data-v-993744c5="" rows="10" class="bg-gray-50 border text-sm rounded-lg block w-full p-2 bg-gray-700 border-gray-600 text-white" required="" value=""></textarea></div>
        </div>
        <div class="h-1/8 w-full flex gap-4 items-center justify-end"><button id="cancelButton" class="bg-gray-700 hover:bg-gray-600 rounded-lg text-lg font-bold p-2 text-white w-1/2 hover:cursor-pointer"> Annuler </button><button id="validateButton" class="bg-green-700 hover:bg-green-600 rounded-lg text-lg font-bold p-2 text-white w-1/2 hover:cursor-pointer disabled:bg-green-600/20 hover:disabled:cursor-not-allowed" disabled=""> Créer </button></div>
      </div>
    </div>
  </div>
</div>"
`;
